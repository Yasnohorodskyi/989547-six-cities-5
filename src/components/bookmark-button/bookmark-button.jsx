import React from "react";
import {connect} from "react-redux";
import {updateOfferBookmarkStatus} from "../../store/api-action.js";
import classNames from "classnames";
import propTypes from "prop-types";

const BookmarkButton = ({type, id, isFavorite, changeBookmarkStatus}) => {
  const {name, width, height} = type;
  const handleClick = () => {
    changeBookmarkStatus(id, Number(!isFavorite));
  };

  const buttonActiveClass = () => {
    if (isFavorite) {
      return `${name}__bookmark-button--active`;
    }
    return ``;
  };

  const svgClass = () => `${name}__bookmark-icon`;

  return (
    <button className={classNames(`${name}__bookmark-button button`, buttonActiveClass())} type="button" onClick={handleClick}>
      <svg className={svgClass()} width={width} height={height}>
        <use xlinkHref="#icon-bookmark"></use>
      </svg>
      <span className="visually-hidden">To bookmarks</span>
    </button>
  );
};

const mapDispatchToProps = (dispatch) => ({
  changeBookmarkStatus(id, status) {
    dispatch(updateOfferBookmarkStatus(id, status));
  }
});

BookmarkButton.propTypes = {
  type: propTypes.shape({
    name: propTypes.string.isRequired,
    width: propTypes.number.isRequired,
    height: propTypes.number.isRequired
  }).isRequired,
  id: propTypes.number.isRequired,
  isFavorite: propTypes.bool.isRequired,
  changeBookmarkStatus: propTypes.func.isRequired
};

export {BookmarkButton};
export default connect(null, mapDispatchToProps)(BookmarkButton);
