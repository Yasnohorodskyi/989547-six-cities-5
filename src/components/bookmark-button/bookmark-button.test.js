import React from "react";
import {BookmarkButton} from "./bookmark-button";
import renderer from "react-test-renderer";

const BookmarkButtonTypesWithProperties = {
  PROPERTY: {
    name: `property`,
    width: 31,
    height: 33
  },
  PLACE_CARD: {
    name: `place-card`,
    width: 18,
    height: 19
  }
};

const id = 45;

const isFavorite = true;

describe(`Should Bookmark button render correctly`, () => {
  it(`Button with active status`, () => {
    const tree = renderer
    .create(<BookmarkButton
      type={BookmarkButtonTypesWithProperties.PROPERTY}
      id={id}
      isFavorite={isFavorite}
      changeBookmarkStatus={() => {}}
    />)
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Button with non active status`, () => {
    const tree = renderer.create(<BookmarkButton
      type={BookmarkButtonTypesWithProperties.PROPERTY}
      id={id}
      isFavorite={!isFavorite}
      changeBookmarkStatus={() => {}}
    />)
    .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
