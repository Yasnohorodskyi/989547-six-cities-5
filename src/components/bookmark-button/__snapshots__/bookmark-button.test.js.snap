// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Bookmark button render correctly Button with active status 1`] = `
<button
  className="property__bookmark-button button property__bookmark-button--active"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height={33}
    width={31}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`Should Bookmark button render correctly Button with non active status 1`] = `
<button
  className="property__bookmark-button button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height={33}
    width={31}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;
